### PR TITLE
Refactor SafeAttributePersistenceProvider and AttributePersistenceProvider to move ember-specific bits to codegendatamodelprovider

### LINKED ISSUE #36472: Decouple ember-specific AttributePersistenceProvider/DefaultAttributePersistenceProvider from libCHIP (only src/app should use it and SafeAttributePersistenceProvider should be separated out)
DefaultAttributePersistenceProvider is ember-specifc (relies on ember metadata and types) and so is DeferredAttributePersistenceProvider.

We have SafeAttributePersistenceProvider which is more generic, however right now that is provided by the ember persistence as well, so logic should be split out.

### PR DESCRIPTION
**Pull request by andy31415**
Intent for this is to reduce ember-specific calls in generic calls, so that we can eventually have applications that have providers without any ember dependencies.

This fixes #36472

### Changes:

- Separate out `SafeAttributePersistenceProvider` as its own stand-alone class. Our unit tests only test this one so move the tests accordingly
- Make AttributePersistenceProvider use composition instead of inheritance on SafeAttributePeristenceProvider
- Make Server be responsible of `SafeAttributePersistenceProvider` (since that is in libCHIP and does not depend on Ember) and make CodegenDataModelProviuder initialize the AttributePersistenceProvider that `attribute-storage.cpp` and a few other clusters use. Recommendation seems to be to use the `SafeAttributePersistenceProvider`

### CODE REVIEW

File: src/BUILD.gn [hunk 1]
```diff
@@ -100,10 +100,7 @@ if (chip_build_tests) {
     }
 
     if (chip_device_platform != "efr32") {
-      tests += [
-        "${chip_root}/src/app/tests",
-        "${chip_root}/src/app/util/persistence/tests",
```
Comment (bzbarsky-apple) [inline src/BUILD.gn:105]: Why this change?  Seems unrelated.

File: src/app/BUILD.gn [hunk 2]
```diff
@@ -390,7 +390,10 @@ source_set("command-handler-impl") {
 }
 
 source_set("attribute-persistence") {
-  sources = [ "SafeAttributePersistenceProvider.h" ]
+  sources = [
+    "SafeAttributePersistenceProvider.cpp",
```
Comment (bzbarsky-apple) [inline src/app/BUILD.gn:394]: Seems completely unrelated.

File: src/app/SafeAttributePersistenceProvider.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
+/*
```
Comment (bzbarsky-apple) [inline src/app/SafeAttributePersistenceProvider.cpp:1]: This looks like a bad merge of https://github.com/project-chip/connectedhomeip/pull/36473 or something?

File: src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp [hunk 4]
```diff
@@ -299,8 +303,37 @@ unsigned FindNextDeviceTypeIndex(Span<const EmberAfDeviceType> types, const Data
...
 
+CHIP_ERROR CodegenDataModelProvider::Startup(DataModel::InteractionModelContext context)
+{
+    ReturnErrorOnFailure(DataModel::Provider::Startup(context));
+
+    // Ember/codegen requires a persistence provider. Ensure one is set.
+    if (GetAttributePersistenceProvider() == nullptr)
+    {
+        ReturnErrorOnFailure(gDefaultAttributePersistenceProvider.Init(GetSafeAttributePersistenceProvider()));
+        SetAttributePersistenceProvider(&gDefaultAttributePersistenceProvider);
```
Comment (bzbarsky-apple) [inline src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:319]: Why is this being handled here?  This seems very much unrelated to just injecting the data model provider...

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
The persistence refactor should be in a separate PR.

File: examples/air-purifier-app/ameba/main/chipinterface.cpp [hunk 5]
```diff
@@ -133,6 +134,7 @@ static void InitServer(intptr_t context)
     // Init ZCL Data Model and CHIP App Server
     static chip::CommonCaseDeviceServerInitParams initParams;
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider = chip::app::CodegenDataModelProviderInstance();
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/air-purifier-app/ameba/main/chipinterface.cpp:137]: These were updated via a sed script, since there are a lot of files updated. Hence the generic path instead of checking every single `using`.

File: src/app/SafeAttributePersistenceProvider.cpp (hunk 3 above)
Comment (andy31415) [reply src/app/SafeAttributePersistenceProvider.cpp:1]: This is a separate update, #36473 did not have a split dependency between the "safe" persistence and the attribute persistence. The safe one is not ember-specific so it can live in `app` where as the "non-safe" one is ember specific.

I flipped the dependencies so that I can have proper ordering in DataModel::Provider initialization code:
  - I want src/app to not depend on ember, so it cannot initialize a "attributepersistence"
  - attributepersistence requires a "safe attribute persistence" (which can be initialized by Server)
  - Moved the codegen provider to initialize the ember side of the persistence provider.

File: examples/air-purifier-app/ameba/main/chipinterface.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/air-purifier-app/ameba/main/chipinterface.cpp]: Updated. From what I see though there are tons of files also doing `using namespace chip` and still having `chip::CommonCaseDeviceServerInitParams`:

image

I am not signing up to fix those...

File: src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp (hunk 4 above)
Comment (andy31415) [reply src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:319]: It is a sideffect of initialization order: data model provider initializes the ember persistence, so I need a re-ordering of when each of the persistence providers are initialized/updated.

I found splitting somewhat harder since this is directly required by me passing in the codegendatamodel to the server init instead of having it default-pulled whenever `InteractionModelEngine::GetDataModeProvider` is called.

I can try to split it out, however would prefer not to due to it already being reasonably tangled.

Comment (bzbarsky-apple) [reply src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:319]: I guess here's my confusion: before this PR, there is no mention of attribute persistence in the codegen data model provider at all.  Is this a hidden dependency that is covered up by accidents of initialization order?  But if the codegen data model provider does not use attribute persistence, why should it be its job to initialize it at all?

Comment (andy31415) [reply src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:319]: I am trying to get a way for applications to not know about ember at all. One of the final places that I found was that `AttributePersistenceProvider` is ember-specific because it uses `EmberAfAttributeMetadata`. So then a `DefaultAttributePersistenceProvider` ties things to ember and the server having such a member in https://github.com/project-chip/connectedhomeip/blob/master/src/app/server/Server.h#L687 make it so that Server is hard-coding ember usage.

My delta looks to completely separate ember-specific parts in `util` and keep non-ember (but generally usable by anyone) parts still in app - this is the `SafeAttributePersistenceProvider`

Comment (bzbarsky-apple) [reply src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:319]: As discussed on Slack:

This PR is conflating two things:

1. The DataModel::Provider should be injectable.
2. The codegen provider calls some functions that for their proper functioning rely on  `SafeAttributePersistenceProvider` having been called.  But they also rely on `emberAfInit` being called and `emberAfEndpointConfigure` being called, at the very least.

If we want to do a PR that makes sure to move all the "init the Ember stack" bits into one place, great.  But that has nothing to do with whether the DataModel::Provider is injectable, and the injectability can be done either before or after the "move to one place" changes.

File: src/app/SafeAttributePersistenceProvider.h [hunk 6]
```diff
@@ -141,7 +154,7 @@ class SafeAttributePersistenceProvider
      * @param [in] aPath the attribute path for the data being written.
      * @param [in] aValue the data to write. The value will be stored as-is.
      */
-    virtual CHIP_ERROR SafeWriteValue(const ConcreteAttributePath & aPath, const ByteSpan & aValue) = 0;
+    virtual CHIP_ERROR SafeWriteValue(const ConcreteAttributePath & aPath, const ByteSpan & aValue);
```
Comment (bzbarsky-apple) [inline src/app/SafeAttributePersistenceProvider.h:157]: No, the whole point of SafeAttributePersistenceProvider is that it's an interface, which can be implemented however is convenient to the application, no?  In particular, it might not be implemented on top of PersistentStorageDelegate at all!  

I'm stopping here, since this seems like a pretty fundamental issue.  So just to be clear:

1. This is meant to be a pure interface.
2. Some specific implementations of this interface might exist, and the SDK should provide a default one to make things easy for people who want to use that default.
3. The implementation to actually use should be injected.  I know that's not what we do now, and Server just hardcodes a `DefaultAttributePersistenceProvider`, but that's a TODO/bug to fix, not something to reify in our API surface.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/SafeAttributePersistenceProvider.h (hunk 6 above)
Comment (andy31415) [reply src/app/SafeAttributePersistenceProvider.h:157]: Following up offline to figure out an approach here. This is what we have today (has a very organic-grown feel to it):

image

**Comment by andy31415** [comment]
Closing for #36658 (after offline discussion about an updated approach)
